### ISSUE TITLE
Pass in correct explicitlySpecified and selectors when creating TestDefinitions 

### ISSUE THREAD

**Issue by antoine-mulet** [sbt/sbt#5609]
Simply something that has never been implemented it seems. For example Scalatest relies on these two fields to implement the tasks run logic so it's a shame they are currently hardcoded.

https://github.com/sbt/sbt/blob/62c6cad487c72b38f67b765b2519b7e1b04560b8/main-actions/src/main/scala/sbt/Tests.scala#L570-L575

This is the first time I take a look at the sbt code but if someone can provide me with some guidance I would be happy to contribute.

**Comment by eed3si9n** [comment]
@antoine-mulet Thanks for the report. Could you provide reproduction steps so we can reproduce your problem on _our computers_?

**Comment by antoine-mulet** [comment]
@eed3si9n sorry if something wasn't clear. I didn't think a proper repro was necessary here because we are talking about an interface that has been partially implemented using hardcoded default values.

To give a bit more details, here is what I understand about the issue:

sbt `TestDefinition`

https://github.com/sbt/sbt/blob/975e3db43ec80110839034b23bdddac5a4614567/testing/src/main/scala/sbt/TestFramework.scala#L82-L87

maps to test-interface `TaskDef`

https://github.com/harrah/test-interface/blob/17a94641942546e21f4c6b300a3360be2d2888f6/src/main/java/sbt/testing/TaskDef.java

In there it is very clear what `explicitlySpecified` is supposed to be

https://github.com/harrah/test-interface/blob/17a94641942546e21f4c6b300a3360be2d2888f6/src/main/java/sbt/testing/TaskDef.java#L15-L43

and same for `selectors`

https://github.com/harrah/test-interface/blob/17a94641942546e21f4c6b300a3360be2d2888f6/src/main/java/sbt/testing/Selector.java

However `explicitlySpecified` and `selectors` are simply not implemented on the sbt side. So down the line when test frameworks like ScalaTest rely on those fields, it leads to bugs like https://github.com/scalatest/scalatest/issues/417 for example.

That being said, to answer the original question, here is how you can reproduce the issue with ScalaTest:

1. Define a Scalatest suite 

```
import org.scalatest.DoNotDiscover
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers

@DoNotDiscover
class FooTestSuite extends AnyFlatSpec with Matchers {

  it should "run" in {
    1 + 1 shouldBe 2
  }
}
```

2. Run it with sbt 

`sbt "testOnly com.foo.test.FooTestSuite"`

3. The output shows that no tests were run

```
[info] ScalaTest
[info] Run completed in 23 minutes, 27 seconds.
[info] Total number of tests run: 0
[info] Suites: completed 0, aborted 0
[info] Tests: succeeded 0, failed 0, canceled 0, ignored 0, pending 0
[info] No tests were executed.
```

And this is because `explicitlySpecified` was set to `false` by sbt (as always in the current implementation) while it should have been set to `true` to let the suite be discovered properly. The same way any logic relying on the correct values to be set for `selectors` would potentially lead to bugs.

Another way to double check the issue is to run the same sbt command in debug mode with breakpoints in the sbt / ScalaTest code to inspect the values of `explicitlySpecified` and `selectors`.

I hope everything is clearer with the explanations above, if still not please let me know.

**Comment by antoine-mulet** [comment]
Hi @eed3si9n, do you have any comments? Is everything above clear?

Thanks,
Antoine

**Comment by eed3si9n** [comment]
Thanks for unpacking this. `@DoNotDiscover` sounds like a good repro step for testing.

**Comment by Eruis2579** [comment]
Hi. @eed3si9n , Can I work on this issue?

**Comment by eed3si9n** [comment]
@Eruis2579 Yes, plz. As per usual check out our [Contributor's Guildeline](https://github.com/sbt/sbt/blob/develop/CONTRIBUTING.md) and:
1. Make sure you can reproduce the problem and the fix manually
2. Include tests for the change
3. Follow our AI assisted contributions guideline (please disclose AI usages in the PR description, if any)
4. Multiple contributors might work on this issue in parallel
5. Generally we recommend getting CI to pass on a forked repo before sending a pull request (pull requests that do not pass tests will be closed after a few days)
